keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,275 @@
+
+### Non-Goals
+
+This proposal does not plan to address how the underlying PVs are treated on PVC deletion. 
+That functionality will continue to be governed by the ReclaimPolicy of the storage class. 
+
+## Proposal
+
+### Background
+
+Controller `garbagecollector` is responsible for ensuring that when a statefulset 
+set is deleted the corresponding pods spawned from the StatefulSet is deleted. 
+The `garbagecollector` uses `OwnerReference` added to the `Pod` by statefulset controller
+to delete the Pod. Similar mechanism is leveraged by this proposal to automatically 
+delete the PVCs created by the StatefulSet controller.
+
+### Changes required
+
+The following changes are required:
+
+1. Add `PersistentVolumeClaimDeletePolicy` entry into StatefulSet spec inorder to make this feature an opt-in.
+2. Provide the following PersistentVolumeClaimPolicies:
+   * `Retain` - this is the default policy and is considered in cases where no policy is specified. This would be the existing behaviour - when a StatefulSet is deleted, no action is taken with
+       respect to the PVCs created by the StatefulSet.
+   * `RemoveOnScaledown` - When a pod is deleted on scale down, the corresponding PVC is deleted as well. 
+       A scale up following a scale down, will wait till old PVC for the removed Pod is deleted and ensure 
+       that the PVC used is a freshly created one.
+   * `RemoveOnStatefulSetDeletion` - PVCs corresponding to the StatefulSet are deleted when StatefulSet
+       themselves get deleted.
+3. Add `patch` to the statefulset controller rbac cluster role for `persistentvolumeclaims`.
+
+### User Stories (optional)
+
+#### Story 1
+User environment is such at the content of the PVCs which are created automatically during StatefulSet 
+creation need not be retained after the StatefulSet is deleted. User also requires that the scale 
+up/down occurs in a fast manner, and leverages any previously existing auto created PVCs within the 
+life time of the StatefulSet. An option needs to be provided for the user to auto-delete the PVCs 
+once the StatefulSet is deleted. 
+
+User would set the `PersistentVolumeClaimDeletePolicy` as `RemoveOnStatefulSetDelete` which would ensure that 
+the PVCs created automatically during the StatefulSet activation is removed once the StatefulSet 
+is deleted.
+
+#### Story 2
+User is cost conscious but at the same time can sustain slower scale up(after a scale down) speeds. Needs 
+a provision where the PVC created for a pod(which is part of the StatefulSet) is removed when the Pod 
+is deleted as part of a scale down. Since the subsequent scale up needs to create fresh PVCs, it will 
+be slower than scale ups relying on existing PVCs(from earlier scale ups). 
+
+User would set the `PersistentVolumeClaimDeletePolicy` as 'RemoveOnScaledown' ensuring PVCs are deleted when corresponding
+Pods are deleted. New Pods created during scale  up followed by a scaledown will wait for freshly created PVCs.
```

Comment:
is there a real world example of such an application  ?if PVC should be deleted when pod is deleted, how is this different than a pod using an emptyDir ? I would vote for some more concrete use cases here

Reply:
The volume is retained if the pod is rescheduled, eg if the node goes down or is upgraded.

+1 to adding the explicitly so it's clear.

Reply:
Or statefulset rolling update

Reply:
Will update the pod rescheduling note shortly.